### DISCUSSION TITLE
Version 3.11.0 retrospective

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#3641] [General]
# Pain points

- It took 7 month since the last version was released. This was initially OK, because there was little progress going on and the project has no release plan until a new version was to be expected. However, then a few regressions were reported while new features were added. This diluted the release scope and a cycle of "let's also wait for that PR" started.
- A long time between releases made the release script unusable - in particular the directory changes broke a lot of code that needed to be adjusted and rewritten.
- The whole release process contains too many moving parts and is far away from a situation where regular releases are possible.
- BJData showed the value of Google's OSS-Fuzz. However, it was a lot of manual work to copy the findings from one (private) issue tracker to GitHub.
- Collecting all changes in a discussion thread was helpful, but writing the final release notes was still a lot of manual work and blocked the release for another 24 hours.
- The CI has gotten too slow for fast development cycles. At the same time, the extensive test suite is needed to avoid regressions.
- Using labels to describe the state of issues and PRs is not really effective.
- The development process is currently too focussed on my machine and my way of working. Collaborating with other active developers was hindered this way.
- Publishing the documentation is currently a manual step and was often forgotten even after extensive documentation improvements.
- Some features were merged "half-ready", and subsequent PRs were needed to polish documentation, tests, etc.

# Possible improvements

- [ ] Release more often.
- [ ] Simplify the release script and remove (more) hard-coded paths and dependencies to implementation details that may change.
- [x] Check which other aspects of the release process can be automated and moved to GitHub actions. ✅ contributor list is now checked automatically, and the whole documentation is now checked as part of the CI #3672
- [x] Write release notes on the fly to avoid a last-minute editing session. ✅ https://github.com/nlohmann/json/discussions/3649
- [ ] Automate creating GitHub issues from OSS-Fuzz issues.
- [ ] Check how the CI can be speed up.
- [ ] Check out GitHub projects to better plan releases.
- [ ] Check how active developers can better be integrated into the project.
- [x] Agree on some kind of communication channel other than GitHub issues/discussions to better synchronize. ✅ It's Discord: https://discord.gg/QCbDsv3X
- [x] Automate publication of documentation when merging to develop. ✅ https://github.com/nlohmann/json/pull/3660

**Comment by falbrechtskirchinger** [comment]
My 2 cents:

Release more often, announce and implement strict feature freezes once development has dragged on for a while.

I've previously mentioned that I had plans for `json-ci` to speed up the initialization phase of the jobs in the Ubuntu workflow. I'll be opening an issue over there soon-ish. Of course, the majority of time is still spent elsewhere.
Maybe we need a different (sub-)set of tests to run on each PR/merge-to-`develop` and have scheduled runs with all tests? I don't know if delaying finding regressions will create more work in the end...

Regarding an external communication channel, I couldn't agree more. Discord is a popular option. Telegram might be another one. IRC is apparently not dead yet. :-)
I personally favor Telegram and am not a fan of Discord – the company, but wouldn't mind either option.

I like your other ideas for improvements as well.

**Reply by gregmarr** [reply]
There is a Slack workspace for this project, though in a month they're changing their free plan from "you can see the last 10,000 messages" to "you can see the last 90 days" so it's not great for any conversation that needs to be available for a significant amount of time.

**Reply by nlohmann** [reply]
I think then it would be Discord. I will set up a server.

**Reply by nlohmann** [reply]
Here it is: https://discord.gg/QCbDsv3X - it currently just has a `#general` channel where some GitHub events are pushed. Let's make this our new home.

**Reply by falbrechtskirchinger** [reply]
See you there.
